enhancements/mco-kargs-day1-support.md

```diff
@@ -0,0 +1,347 @@
+- Update the MachineConfigDaemon to compare k-args set on MachineConfigs with
+  those set via rpm-ostree to determine if there are changes requiring a reboot.
+
+Pros:
+- Requires no additional packaging in RHCOS or other tricky methods to update
+  what's available on the host.
+- Allows new nodes in clusters with 4.2 bootimages to have k-args set without
+  the additional reboot.
+- Brings the MCD closer to the kubernetes model of inspecting current state and
+  reconciling desired spec with that.
+
+Cons:
+- Further special-cases kernel args as state passed to pivot. Any other config
+  that should be set before os pivot would need to be specially handled as well.
+- Determining the current state from the host rather than the current config
+  could get tricky, and is different from how other config fields are handled.
+  See https://github.com/openshift/machine-config-operator/pull/245
+
+
+##### Method 2 - MCD-as-pivot packaged in RHCOS
+
+- Installer in 4.2 ships machine-config-daemon package in RHCOS host which handles [early pivot](https://github.com/openshift/machine-config-operator/pull/859).
+- [MCS inverts MachineConfig](https://github.com/openshift/machine-config-operator/pull/868) containing non-ignition part (like osImageURL,kernelArguments) during first boot in /etc/ignition-machine-config-encapsulated.json on hosts
+- During the first boot, MCD is patched to parse the ignition-machine-config-encapsulated.json available on the host to read and set the k-args if specified.
+- machine-config-dameon package in RHCOS is updated so that it has required changes on the host during early boot.
+- MCS is updated so that MCD first boot service runs to process kargs
+- As a result, we get kargs applied on RHCOS nodes during first boot without incurring extra reboot.
+
+Pros:
+- Moves us to a more flexible model where the MCD can be the component managing
+  the host.
+- Removes any drift between how k-args are handled on day-1 or day-2 by using
+  the same code paths.
+
+Cons:
+- Requires latest MCD package with necessary fixes in RHCOS shipped with installer, any further changes needs [updating bootimage](https://github.com/openshift/os/issues/381) which is currently not in place.
+- Will only work on clusters with bootimages at 4.3 or higher, i.e. clusters
+  installed at 4.1 or 4.2 and upgraded will not have k-args applied without a
+  double reboot on new nodes spun up.
+- Running the MCD as a binary removes all the advantages we get from containers.
+
+
+##### Method 3 - MCD-as-pivot fed in through ignition
+
+- Same as method 2 except rather than packaging the MCD in RHCOS, feed it to the
+  hosts via their ignition file.
+- The MCS would need to extract the MCD through the os-container place it in the
+  ignition files it serves.
+
+Pros:
+- Same as method 2
+- Works for all versions of bootimages, including 4.1 and 4.2
+
+Cons:
+- Significantly more complexity in the MCS
+- Additional work to fit into a tight release schedule
```

Comment:
if we find an upgrade path which isn't disruptive I think we can move to Method 3 for >4.3

Reply:
ok, that's outlined in https://github.com/openshift/enhancements/pull/37/files#diff-d6b1c30906b6eb5949f027b5ed336362R156 :+1: